docker_push_latest_if_changed.py

```diff
@@ -1,5 +1,182 @@
 #!/usr/bin/env python3
 import argparse
+import hashlib
+import logging
+from collections import namedtuple
+
+import docker
+
+
+ImageParityFields = namedtuple(
+    'ImageParityFields',
+    ['image_cmds_hash', 'packages_hash'],
+)
+
+GET_PACKAGE_LIST_COMMAND = 'dpkg -l'
+
+
+def _docker_push_latest_if_changed(source, target, is_dry_run):
+    docker_client = _get_docker_client()
+    _validate_source(source, docker_client)
+    target = _get_sanitized_target(source, target)
+
+    logging.info('Pushing source image')
+    _push_image(source, docker_client, is_dry_run)
+    try:
+        logging.warning('Pulling target image')
+        docker_client.pull(target)
+    except docker.errors.NotFound:
+        logging.warning((
+            'Target image {} was not found in the registry. '
+            'Going to attempt to tag and push the target image anyway.'
+        ).format(target))
+        _tag_image(source, target, docker_client, is_dry_run)
+        _push_image(target, docker_client, is_dry_run)
+    else:
+        if _is_image_changed(source, target, docker_client):
+            logging.warning((
+                'Source image was found to be different from the current '
+                'version of the target image. Tagging and pushing new target '
+                'image.'
+            ))
+            _tag_image(source, target, docker_client, is_dry_run)
+            _push_image(target, docker_client, is_dry_run)
+        else:
+            logging.warning((
+                'Source image was found to be the same as the current '
+                'version of the target image. Not pushing the target image.'
+            ))
+
+
+def _validate_source(source, docker_client):
```

Comment:
can't this function be replaced with `assert ':' in source, source`?

Reply:
Unfortunately the registry portion of a tag can have a colon as well (i.e. 127.0.0.1:12345/repo:latest), so I can't just do that. I didn't want to use a regex and play hide and seek with edge cases, so I figured that I might as well use something definitive like checking the inspect contents. Plus inspect has the added benefit of verifying that the image exists in the first place.

Reply:
ah! in that case just use `urllib.parse.urlparse` -- they _are urls_ after all :)

this'll remove the ambiguity about port I believe and give you some confidence on whether they're tagged or not